README.md
# 多标签分类

## 算法网络结构

使用tensorflow自带MobileNetV2作为主干网络，最后使用GAP代替全连接进行输出，输出层激活函数为sigmoid，使用二进制交叉熵作为loss，二进制准确度作为监督。

## 环境

python >=3.8.7

```shell
pip install requirement.txt
```

## 训练

入参

| 参数名称          | 用途                 | 默认值   |
| ------------- | ------------------ | ----- |
| batch_size    | 规定训练的batch size    | 32    |
| epochs        | 训练多少个epochs        | 10    |
| learning_rate | 学习率                | 0.001 |
| gpu           | 使用那块卡进行训练          | 0     |
| dataset_id    | data文件夹下数据集放在哪个id下 | 0     |

执行命令

```shell
python main.py
```


history/history.txt
{'loss': [0.2195880115032196, 0.1930185705423355, 0.17683327198028564, 0.17135512828826904, 0.17135512828826904, 0.17135512828826904], 'binary_accuracy': [0.9034140110015869, 0.9148392677307129, 0.9219666123390198, 0.924480140209198, 0.924480140209198, 0.924480140209198], 'val_loss': [0.28747254610061646, 0.19651398062705994, 0.17968446016311646, 0.17971843481063843, 0.17971928417682648, 0.17971929907798767], 'val_binary_accuracy': [0.888903796672821, 0.9130608439445496, 0.9203975200653076, 0.9203352332115173, 0.9203346967697144, 0.9203346967697144]}

test.py
import tensorflow as tf

model = tf.keras.models.load_model('./savedmodel')

image = tf.keras.utils.load_img('./data/test/test3.jpeg', target_size=(128, 128))
image_array = tf.keras.utils.img_to_array(image)
image_array = tf.expand_dims(image_array, 0) # Create a batch

predictions = model.predict(image_array)
print(predictions)

REST_test.py
import time
import json
import requests

import tensorflow as tf
image = tf.keras.utils.load_img(
    './data/test/test3.jpeg', target_size=(128, 128))
image_array = tf.keras.utils.img_to_array(image)
image_array = tf.expand_dims(image_array, 0)  # Create a batch
with open('test.json', 'w') as f:
    f.write(json.dumps({'inputs': image_array.numpy().tolist()}))
    f.close()

start = time.time()
r = requests.post("http://192.168.113.63:8507/v1/models/face_attribute:predict",
                  data=json.dumps({'inputs': image_array.numpy().tolist()}),
                  headers={'Content-Type': 'application/json'})
end = time.time()

print("r=", r.text)
print('time:', end - start)

model.py
import tensorflow as tf
from tensorflow.keras.applications import mobilenet_v2
from tensorflow import keras
from tensorflow.keras.layers import Input
from tensorflow.keras import Model

def Backbone(num_classes=None):
    """
    This function creates the MobileNetV2 backbone.
    """
    def backbone(x):
        extractor = mobilenet_v2.MobileNetV2(input_shape=x.shape[1:], include_top=False, weights='imagenet', classes=num_classes)
        preprocess = tf.keras.applications.mobilenet_v2.preprocess_input
        return Model(inputs=extractor.input, outputs=extractor.output, name='Backbone')(preprocess(x))
    return backbone

def MultiLabelClassifier(cfg, num_classes, training=False):
    """
    This function creates the MultiLabelClassifier.
    """
    input_size = int(cfg['train']['input_size']) if training else None
    x = inputs = Input([input_size, input_size, 3], name='input')
    x = Backbone(num_classes)(x)
    x = keras.layers.GlobalAveragePooling2D()(x)
    outputs = keras.layers.Dense(num_classes, activation='sigmoid')(x)
    return Model(inputs, outputs, name='MultiLabelClassifier')

dataset.py
import os
import io
from tensorflow.python.data.ops.dataset_ops import AUTOTUNE
import tqdm
import pandas as pd
import numpy as np
import PIL.Image as Image
import tensorflow as tf
from sklearn.model_selection import train_test_split

class LoadData(object):
    def __init__(self, image_path, label_path, image_size):
        self.image_path = image_path
        self.label_path = label_path
        self.image_size = image_size

    def extract_data(self, file_path):
        data = pd.read_csv(file_path)
        head = list(data.columns)
        return head, data

    def read_image(self, file_path):
        image = tf.io.read_file(file_path)
        image = tf.image.decode_jpeg(image, channels=3)
        image = tf.image.resize(image, [self.image_size, self.image_size])
        return image

    def image_verify(self, image_path):
        try:
            image = Image.open(image_path)
            image.verify()
            return True
        except:
            return False
        
    def load_data(self):
        ds_head, ds_data = self.extract_data(self.label_path)
        file_list = ds_data[ds_head[0]].tolist()
        curpath = os.path.dirname(os.path.realpath(__file__))
        file_path_list = [os.path.join(curpath, self.image_path, f) for f in file_list]
        # verify images
        print('>'*5, "Verifying images:")
        for image_path in tqdm.tqdm(file_path_list):
            if not self.image_verify(image_path):
                print("{} is not a valid image".format(image_path))
        # Get the classes number
        num_classes = len(ds_head[1:])
        labels = ds_data[ds_head[1:]].values

        # Split the dataset into training and testing sets
        train_file_path_list, val_file_path_list, train_labels, val_labels = train_test_split(file_path_list, labels, test_size=0.2, random_state=42)
        # Create a dataset containing the filenames of the training images
        print('>'*5, 'The train file length is: ', len(train_file_path_list))
        print('>'*5, 'The test file length is: ', len(val_file_path_list))
        # Set all negative labels to 0
        train_labels[train_labels == -1] = 0
        val_labels[val_labels == -1] = 0
        # generate image dataset and label dataset respectively
        train_path_ds = tf.data.Dataset.from_tensor_slices(train_file_path_list)
        train_label_ds = tf.data.Dataset.from_tensor_slices(tf.cast(train_labels, tf.int64))
        train_ds = train_path_ds.map(self.read_image, num_parallel_calls=AUTOTUNE)

        val_path_ds = tf.data.Dataset.from_tensor_slices(val_file_path_list)
        val_label_ds = tf.data.Dataset.from_tensor_slices(tf.cast(val_labels, tf.int64))
        val_ds = val_path_ds.map(self.read_image, num_parallel_calls=AUTOTUNE)
        
        # print(train_label_ds.take(1).as_numpy_iterator().next())
        # zip the image and labels together
        train_dataset = tf.data.Dataset.zip((train_ds, train_label_ds))
        val_dataset = tf.data.Dataset.zip((val_ds, val_label_ds))
        
        return train_dataset, val_dataset, num_classes

main.py
import os
import datetime
import configparser
import tensorflow as tf

from absl import app, flags, logging
from absl.flags import FLAGS
from tensorflow import keras
from tensorflow.keras.callbacks import ModelCheckpoint
from model import MultiLabelClassifier
from dataset import LoadData

flags.DEFINE_string('batch_size', '32', 'batch size')
flags.DEFINE_string('epochs', '10', 'epochs')
flags.DEFINE_string('learning_rate', '0.001', 'learning rate')
flags.DEFINE_string('gpu', '0', 'gpu')
flags.DEFINE_string('dataset_id', '0', 'dataset id')

def modify_config(cfg):
    """
    This function modifies the config.
    """
    cfg['train']['batch_size'] = FLAGS.batch_size
    cfg['train']['epochs'] = FLAGS.epochs
    cfg['train']['learning_rate'] = FLAGS.learning_rate
    return cfg

def main(_):
    """
    This function is the main function.
    """
    # Set the GPU to use
    os.environ['CUDA_VISIBLE_DEVICES'] = FLAGS.gpu
    # Set the batch size
    batch_size = int(FLAGS.batch_size)
    # Set the epochs
    epochs = int(FLAGS.epochs)
    # Set the learning rate
    learning_rate = float(FLAGS.learning_rate)
    # Set the dataset id
    dataset_id = FLAGS.dataset_id
    # Change config file
    curpath = os.path.dirname(os.path.realpath(__file__))
    cfgpath = os.path.join(curpath, "configs", "configs.ini")
    configs = configparser.ConfigParser()
    configs.read(cfgpath, encoding="utf-8")
    # import data
    dataset_path = os.path.join(curpath, "data", dataset_id)
    dataset_loader = LoadData(os.path.join(dataset_path, "images"), 
                                os.path.join(dataset_path, "labels", "list_attr_celeba.csv"),
                                int(configs['train']['input_size']))
    train_ds, val_ds, num_classes = dataset_loader.load_data()
    train_ds = train_ds.batch(batch_size)
    val_ds = val_ds.batch(batch_size)
    # Set model
    model = MultiLabelClassifier(configs, num_classes, training=True)

    cfg = modify_config(configs)
    # Set Checkpoint callback
    checkpoint_path = "./checkpoint/cp-{epoch:04d}.ckpt"
    checkpoint = ModelCheckpoint(checkpoint_path,
                                verbose=1,
                                save_weights_only=True,
                                save_freq=5000)
    # Set tensorboard callback function
    log_dir = "logs/" + datetime.datetime.now().strftime("%Y%m%d-%H%M%S")
    if not os.path.exists(log_dir):
        os.makedirs(log_dir)
    tensorboard_callback = tf.keras.callbacks.TensorBoard(log_dir=log_dir, histogram_freq=1)
    # Optimizer
    optimizer = keras.optimizers.Adam(learning_rate=keras.optimizers.schedules.CosineDecay(learning_rate, 3*len(train_ds)))
    # Compile the model
    model.compile(optimizer=optimizer,
              loss=tf.keras.losses.BinaryCrossentropy(from_logits=False),
              metrics=[tf.keras.metrics.BinaryAccuracy()],
              run_eagerly=True)
    # Model summary
    model.summary()
    # Train the model
    history = model.fit(
        train_ds,   # training data
        validation_data=val_ds,   # validation data
        epochs=epochs,   # number of epochs
        verbose=1,   # verbosity
        callbacks=[tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=5), checkpoint, tensorboard_callback]
    )
    # Save the history
    with open(os.path.join(curpath, "history", "history.txt"), "w") as f:
        f.write(str(history.history))
    
    # Save the model
    model_save = MultiLabelClassifier(configs, num_classes, training=False)
    model_save.load_weights('checkpoint/cp-0006.ckpt')
    model_save.save(os.path.join(curpath, "savedmodel"))

if __name__ == '__main__':
    app.run(main)